### ISSUE TITLE
Using the best identified UQ method for new data

### ISSUE THREAD

**Issue by nhadler** [Novartis/UNIQUE#5]
Thanks for getting the documentation up! I was able to try it out with one of our datasets. What's unclear to me is how I can then use the identified method saved under errors models to then predict the uncertainty of new examples.

For example, unique determined that `UniqueRandomForestRegressor[UQmetrics+predictions](l1)` is the most performant method across the three metrics for my dataset. Does the package then provide any way to use that method for determining the uncertainty of new predictions?

I tried to unpickle this file `error_models/UQmetrics_predictions/UniqueRandomForestRegressor_model.pkl` but it looks like this is just a dataframe of the predictions for each model rather than something I could pass new examples to.

**Comment by gscebba** [comment]
Hi @nhadler ,
at the moment, the package does not offer that feature but we are working on it.

**Comment by mtdhuynh** [comment]
Hi @nhadler,

Thank you for your interest in `UNIQUE`! As mentioned by @gscebba, the "inference" module is still under development, but once released it will allow users to apply the best performing UQ method to new predictions at inference time. Stay tuned for updates!

Re: your issue with the error model - could you please provide a reproducible MWE/code snippet/data, or at least details on the environment and `UNIQUE` version that is causing you the issue? So that we can try to pinpoint the possible bug.

I just re-run the two public examples we provide in the documentation using the latest `UNIQUE` release (`v0.2.1`), and I was able to successfully load the pickled error models as `sklearn` models:

![image](https://github.com/user-attachments/assets/685f2434-d20c-48a1-88c7-ffae27cc7647)

> I tried to unpickle this file error_models/UQmetrics_predictions/UniqueRandomForestRegressor_model.pkl but it looks like this is just a dataframe of the predictions for each model rather than something I could pass new examples to.

Just for completeness: along with the pickled error model, in each error model's subfolder we save the data used to generate (train and validate) the error model themselves, as `csv` files (`data.csv`):

![image](https://github.com/user-attachments/assets/7ab9d6c9-99a8-4340-ab7c-a3427ce4cb38)

Let me know if we can provide any more help!

**Comment by nhadler** [comment]
Hi @mtdhuynh

Sure! Here's an example of what happens when I unpickle the model:
<img width="1107" alt="Screenshot 2024-09-27 at 9 46 23 AM" src="https://github.com/user-attachments/assets/7a1d400d-e951-4007-a82c-65469befbc5d">

I've attached a zipped file containing config file, pickled input data, and outputs. I ran unique in a new conda environment with just python 3.11 installed. One thing that may be different from your examples is that my dataset is only about ~420 entries. 
[UQ.zip](https://github.com/user-attachments/files/17167282/UQ.zip)

**Comment by mtdhuynh** [comment]
@nhadler: I tried to reproduce your issues with a fresh `conda` env with `python=3.11.10` and `unique=0.2.1` using your same data and code you provided, but the error model loads correctly for me. I tried as well with both `unique=0.2.0` and building from source again, with the same end result.

> Sure! Here's an example of what happens when I unpickle the model: <img alt="Screenshot 2024-09-27 at 9 46 23 AM" width="1107" src="https://private-user-images.githubusercontent.com/3527739/371615327-7a1d400d-e951-4007-a82c-65469befbc5d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc1MDg0MjUsIm5iZiI6MTcyNzUwODEyNSwicGF0aCI6Ii8zNTI3NzM5LzM3MTYxNTMyNy03YTFkNDAwZC1lOTUxLTQwMDctYTgyYy02NTQ2OWJlZmJjNWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjhUMDcyMjA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmU5NzRiNjA1YWMwNTUzODlmMzZmMzg0N2E2OGY4YmQ5OTE4YTkwYmY5ZGVlYWJiNzdlMzc0MzRkZTA0ZDg5ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.IdyDTYl5_YazcSrsrRvQaY354OiflbfyEmJ8sqEmXU4">

What you are seeing here is the first output from `pipeline.fit()`, a dictionary with the UQ metrics name and values computed for your input dataset.

Have you tried running the pipeline again from scratch (e.g., deleting the `outputs` folder)?

**Comment by nhadler** [comment]
> @nhadler: I tried to reproduce your issues with a fresh `conda` env with `python=3.11.10` and `unique=0.2.1` using your same data and code you provided, but the error model loads correctly for me. I tried as well with both `unique=0.2.0` and building from source again, with the same end result.
> 
> > Sure! Here's an example of what happens when I unpickle the model: <img alt="Screenshot 2024-09-27 at 9 46 23 AM" width="1107" src="https://private-user-images.githubusercontent.com/3527739/371615327-7a1d400d-e951-4007-a82c-65469befbc5d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc1MDg0MjUsIm5iZiI6MTcyNzUwODEyNSwicGF0aCI6Ii8zNTI3NzM5LzM3MTYxNTMyNy03YTFkNDAwZC1lOTUxLTQwMDctYTgyYy02NTQ2OWJlZmJjNWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MjhUMDcyMjA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmU5NzRiNjA1YWMwNTUzODlmMzZmMzg0N2E2OGY4YmQ5OTE4YTkwYmY5ZGVlYWJiNzdlMzc0MzRkZTA0ZDg5ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.IdyDTYl5_YazcSrsrRvQaY354OiflbfyEmJ8sqEmXU4">
> 
> What you are seeing here is the first output from `pipeline.fit()`, a dictionary with the UQ metrics name and values computed for your input dataset.
> 
> Have you tried running the pipeline again from scratch (e.g., deleting the `outputs` folder)?

I just deleted the outputs and reran the pipeline and now I'm getting the model being pickled instead of the outputs.

**Comment by mtdhuynh** [comment]
Glad to hear the issue was resolved, @nhadler!

--- 

> For example, unique determined that UniqueRandomForestRegressor[UQmetrics+predictions](l1) is the most performant method across the three metrics for my dataset. Does the package then provide any way to use that method for determining the uncertainty of new predictions?

Regarding reproducing at inference time your specific use-case while we develop the official inference module ourselves:

- As you mentioned, the most fitting UQ metric for your dataset is the RF error model that uses as inputs the UQ metrics values and the original predictions, and predicts the L1 prediction error (`UniqueRandomForestRegressor[UQmetrics+predictions](l1)`).
- In order to apply the error model at inference time, you need to reproduce the inputs to it for the new datapoints. In this case, the inputs are: 1) the prediction from your original model, 2) all the input _base_ UQ metrics. In your specific case, the UQ metrics that `UNIQUE` computes (based on the config file you provided) are: 
    - **Distance-based metrics**: you need to provide also the training data in order to compute how distant the new datapoints are from the original training set.
    - **Ensemble variance**: in addition to the point-wise single-value prediction (e.g., the average of the predictions of your ensemble, column `Predictions` in your dataset), you need to also provide the individual model's predictions of your ensemble (column `Ensemble Predictions` in your dataset).
- Finally, you can apply the `sklearn` trained model at inference time.

Below, an example code snippet to do so:

```python
import numpy as np
import pandas as pd
import pickle

from unique.input_type import FeaturesInputType, ModelInputType

# Load original data
with open("<PATH_TO_PKL>/UQ_input.pkl", "rb") as f:
    original_data = pickle.load(f)

# Filter only the training data
training_data = original_data.loc[original_data["Subset"]=="TRAIN"]
# Drop unnecessary/unavailable columns at inference time
training_data = training_data.drop(columns=["Labels", "index"])

# Generate a random datapoint
rng = np.random.default_rng(42)
test_datapoint = {
    "Descriptors": [rng.uniform(-1, 1, size=(79,))],
    "Ensemble Predictions": [rng.uniform(0, 5, size=(4,))],
    "Subset": ["INFERENCE"],
}
test_datapoint["Predictions"] = [np.mean(test_datapoint["Ensemble Predictions"])]
test_datapoint["Ensemble Variances"] = [np.var(test_datapoint["Ensemble Predictions"])]

# Concat together
data = pd.concat([training_data, pd.DataFrame.from_dict(test_datapoint)], axis=0).reset_index(drop=True)

# Initialize UNIQUE inputs
data_inputs = FeaturesInputType(
    column_name = "Descriptors",
    input_values = data["Descriptors"],
    which_set = data["Subset"].to_numpy(),
    problem_type = "regression",
)
model_inputs = ModelInputType(
    column_name = "Ensemble Predictions",
    input_values = data["Ensemble Predictions"],
    which_set = data["Subset"].to_numpy(),
    problem_type = "regression",
)

# Compute UQ values
uq_values = {}
uq_values.update(data_inputs.fit())
uq_values.update(model_inputs.fit())

# Only select the test datapoint's UQ values
test_uq = []
for v in uq_values.values():
    test_uq.append(v[-1]) # In this specific case it's the last element

# Also append the original model's point-wise prediction
test_uq.append(test_datapoint["Predictions"][0])

# Load the trained error model
with open("<PATH_TO_ERROR_MODEL>/outputs/error_models/UQmetrics_predictions/UniqueRandomForestRegressor_model.pkl", "rb") as f:
    trained_error_model = pickle.load(f)

# Predict on the new datapoint
output = trained_error_model.predict([test_uq]) # input shape needs to be (n_samples, n_features)
```

`output` here is the predicted error from your trained error model (in this case, the predicted **L1** error). In your case, you would need to replace the `test_datapoint` section with your actual data, instead of generating a random one.

> [!NOTE]
> This example applies specifically to your provided data and pipeline's output (i.e., with the RF error model trained on UQ metrics and original predictions only being the best performing UQ metric).

If you want to use other UQ metrics and input features in your error model, you will need to compute and add them using a similar workflow as the one above (e.g., _transformed_ UQ metrics). Please, refer to the [docs](https://opensource.nibr.com/UNIQUE/indepth/error_models.html) for more details on the input data to error models and feel free to ask for help!

Similarly, if you wish to show other UQ metrics at inference time, you can follow the same steps as above.

Hope this was helpful! 

Stay tuned for more updates on the inference module and let us know if you have any more issues.

**Comment by nhadler** [comment]
> Glad to hear the issue was resolved, @nhadler!

This looks great! Thanks for taking the time to send a snippet on how to do this---I'll be testing this out immediately. I really appreciate this new tool--we've been wanting to apply more advanced UQ metrics in our lab but not really sure where to start. Thanks!
